Solve issue #282: Update FSSpecFileLister, IoPathFileLister

ykabusalah opened the pull request:
Fixes #282 

### Summary

Updated FSSpecFileLister and IoPathFileLister to optionally take a list of strings as root paths. As per the contributors guideline, added tests to the testsuite to cover changes and ran the pre-commit format checker.

### Changes
- Updated FSSpecFileLister to take either a str or sequence of strings as root path argument (in \_init_ arguments)

- Updated \_init_ of FSSpecFileLister to check if root argument is string or list and make into IterDataPipe

- Updated \_iter_ of FSSPecFileLister to look through all root paths when yielding files

- Added unit test in test_fsspec.py to test for multiple root path input to FSSpecFileLister

- Updated IoPathFileLister to take either a str or sequence of strings as root path argument (in \_init_ arguments)

- Updated \_init_ of IoPathFileLister to check if root argument is string or list and make into IterDataPipe

- Updated \_iter_ of IoPathFileLister to look through all root paths when yielding files

- Added unit test in test\_local_io.py to test for multiple root path input to IoPathFileLister

ykabusalah commented:
@NivekT would you mind taking a look and seeing if this works? just really want to see if there's any extra work left for me to ensure it works fully. thank you!

ykabusalah commented:
@NivekT one test failed for macos build but I see that it has failed on previous pull requests as well. Let me know if there is anything I need to do for this.

Thanks.

NivekT commented:
> @NivekT one test failed for macos build but I see that it has failed on previous pull requests as well. Let me know if there is anything I need to do for this.
> 
> Thanks.

Don't worry about that one. I will have a look at the rest of your PR soon. Thanks!

ykabusalah commented:
> > @NivekT one test failed for macos build but I see that it has failed on previous pull requests as well. Let me know if there is anything I need to do for this.
> > Thanks.
> 
> Don't worry about that one. I will have a look at the rest of your PR soon. Thanks!

Awesome. Sounds good. Any chance you'd be able to take a look at it by tomorrow?

ejguan commented on torchdata/datapipes/iter/load/fsspec.py:
```diff
@@ -49,37 +49,45 @@ class FSSpecFileListerIterDataPipe(IterDataPipe[str]):
-                        yield fs.protocol + "://" + abs_path
+        for root in self.datapipe:
+            fs, path = fsspec.core.url_to_fs(root)
+            is_local = fs.protocol == "file" or not root.startswith(fs.protocol)
```
Do you mind solving https://github.com/pytorch/data/issues/348 as well?
You can basically check if `fs.protocol` is a list. If it is, we can run `any(root.startswith(protocol) for protocol in fs.protocol)`

ejguan commented on torchdata/datapipes/iter/load/fsspec.py:
```diff
@@ -49,37 +49,45 @@ class FSSpecFileListerIterDataPipe(IterDataPipe[str]):
+                        continue
+
+                    # ensure the file name has the full fsspec protocol path
+                    if file_name.startswith(fs.protocol):
```
Ditto

ejguan commented on torchdata/datapipes/iter/load/fsspec.py:
```diff
@@ -49,37 +49,45 @@ class FSSpecFileListerIterDataPipe(IterDataPipe[str]):
+                        else:
+                            abs_path = posixpath.join(path, file_name)
+
+                        if root.startswith(fs.protocol):
+                            yield fs.protocol + "://" + abs_path
```
The logic here might need to be changed as well.

ykabusalah replied:
In what way should this be changed?

NivekT replied:
I believe @ejguan is referring to his other comment here:
https://github.com/pytorch/data/pull/383#discussion_r866189401

It will be great if you can solve #348 as well in this PR, and when you do, the logic here will have to change as well.

ejguan replied:
```suggestion
                        if isinstance(fs.protocol, list):
                            for protocol in fs.protocol:
                                if root.startswith(protocol):
                                   yield protocol + "://" + abs_path
                                   break
                        elif root.startswith(fs.protocol):
                            yield fs.protocol + "://" + abs_path
```

ejguan reviewed (commented):
Thank you for adding this patch.

NivekT commented on test/test_fsspec.py:
```diff
@@ -57,6 +64,26 @@ def test_fsspec_file_lister_iterdatapipe(self):
+    def test_fsspec_file_lister_iterdatapipe_with_list(self):
+        datapipe = FSSpecFileLister(root=["file://" + self.temp_sub_dir.name, "file://" + self.temp_sub_dir_2.name])
+
+        # check all file paths within sub_folder are listed
```
nit: might be simpler to make two sorted `list` and compare them?

NivekT commented on test/test_local_io.py:
```diff
@@ -601,6 +608,17 @@ def test_io_path_file_lister_iterdatapipe(self):
+    def test_io_path_file_lister_iterdatapipe_with_list(self):
+        datapipe = IoPathFileLister(root=[self.temp_sub_dir.name, self.temp_sub_dir_2.name])
+
+        # check all file paths within sub_folder are listed
```
nit: similar to my other comment about comparing sorted list instead

NivekT reviewed (commented):
Thanks for submitting the PR! Just a few small comments.

ykabusalah commented:
@NivekT @ejguan 

This fixes issue #348 and updates Testing Logic



### Summary



- Updated FSSpecFileLister to check if protocol for root is a string or list. S3 returns a list, so this handles that return and updates logic.

- Updated IO and FSSpec unit tests to sort and compare lists rather than looping and checking



### Changes



- Updated FSSpecFileLister to check if fs protocol is a list or a string

- If it is a string, it puts it into a list and handles the rest of the logic of the FileLister as if it were a list

- Uses Python's "any()" functionality to check conditions against the protocol list

- Check the beginning of the root against all protocols and returns if it is a match, else returns the absolute path

- Updated IO and FSSpec unit tests to sort and compare lists rather than looping and checking

ejguan commented on torchdata/datapipes/iter/load/fsspec.py:
```diff
@@ -49,37 +49,55 @@ class FSSpecFileListerIterDataPipe(IterDataPipe[str]):
+            else:
+                protocol_list = fs.protocol
+
+            is_local = fs.protocol == "file" or not any(root.startswith(protocol) for protocol in fs.protocol)
```
you mean `protocol_list`, right?

ejguan commented on torchdata/datapipes/iter/load/fsspec.py:
```diff
@@ -49,37 +49,55 @@ class FSSpecFileListerIterDataPipe(IterDataPipe[str]):
+
+                        starts_with = False
+                        for protocol in protocol_list:
+                            if root.startswith(protocol):
+                                starts_with = True
+                                yield protocol + "://" + abs_path
```
Let's add a `break` when the root path starts with the protocol the first time. Otherwise, there can be multiple paths yielded.

ejguan reviewed (commented):
Look good to me in general. Two more comments below.

ykabusalah commented:
@ejguan @NivekT 

This new commit should address the remaining outstanding concerns: added a variable change to make it congruent with the rest of the changes as well as a break in a for loop to stop yielding values when we found the one we wanted. 

Thanks.

NivekT reviewed (approved):
LGTM! Thanks for working on this PR!

ejguan commented on torchdata/datapipes/iter/load/fsspec.py:
```diff
@@ -49,37 +49,56 @@ class FSSpecFileListerIterDataPipe(IterDataPipe[str]):
+                root,
+            ]
+        if not isinstance(root, IterDataPipe):
+            self.datapipe: Union[IterDataPipe, IterableWrapper] = IterableWrapper(root)
```
nit: `IterDataPipe` should be sufficient as the type hint.

NivekT replied:
Fixed internally

ejguan commented on torchdata/datapipes/iter/load/iopath.py:
```diff
@@ -72,20 +72,28 @@ def __init__(
+                root,
+            ]
+        if not isinstance(root, IterDataPipe):
+            self.datapipe: Union[IterDataPipe, IterableWrapper] = IterableWrapper(root)
```
Ditto

NivekT replied:
Fixed internally
